xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -400,8 +410,45 @@ public String toString() {
           }
         }
 
+        class BackendMetricsRecordPicker extends SubchannelPicker {
+          private final SubchannelPicker delegate;
+
+          private BackendMetricsRecordPicker(SubchannelPicker delegate) {
+            this.delegate = delegate;
+          }
+
+          @Override
+          public PickResult pickSubchannel(PickSubchannelArgs args) {
+            PickResult result = delegate.pickSubchannel(args);
+            if (!result.getStatus().isOk()) {
+              return result;
+            }
+            if (result.getSubchannel() == null) {
+              return result;
+            }
+            ClientLoadCounter localityCounter = statsStore.getLocalityCounter(locality);
+            if (localityCounter == null) {
```

Comment:
Given the current implementation, this should not be possible. If it did happen, it would probably due to a race, either overlooked by us now or introduced in the future, between removal of the locality and the picking, which would be a very short window thus have negligible impact. Since we have tests to cover the non-racy cases, it should be fine to tolerate the racy cases. I suggest keeping the null check and removing the log. The same goes for `XdsLoadStatsStore#interceptPickResult()`.

Further thinking about it, do we have to call `statsStore.getLocalityCounter()` for each pick? The locality is an invariant to the picker, why can't it hold a `ClientLocalCounter`? Then there shouldn't be any race. Similarly, `StatsStore#interceptPickResult(PickResult, XdsLocality)` could be `StatsStore#interceptPicker(SubchannelPicker, XdsLocaliy)` instead.

Reply:
Another way I can think about is to completely decouple `StatsStore` with `SubchannelPicker` (or `PickResult`) just like what an [`LocalityMetricsListener`](https://github.com/grpc/grpc-java/blob/f079a9b570363437c03c7514d24a86eb92563feb/xds/src/main/java/io/grpc/xds/ClientLoadCounter.java#L296) does. That is, we do not implement any `StatsStore#interceptPickResult()` (or `StatsStore#interceptPicker()`) in `StatsStore`'s implementation. Instead, we provide a class (say, `LoadRecordingSubchannelPicker`) that behaves like a `ForwardingSubchannelPicker`, so the user (i.e., xDS lb's `LocalityStore`) can just wrap its picker by constructing such a `LoadRecordingSubchannelPicker`. In this way, `StatsStore` has a very clear interface that only does add/get/remove/ counters and record drops. And any further changes made on how we wrap `SubchannelPicker` or `PickResult` would not affect `StatsStore` (it's just an in-memory storage for load).



PS: I remember long time ago you mentioned the reason we had `StatsStore#interceptPickResult()`was to force user to not forget doing the interception. Well, I don't think that's a strong argument. Users SHOULD know they need to apply load-recording logic to subchannels picked, then our `LoadRecordingSubchannelPicker` is there to help.

Reply:
This is what comes up to my mind:

```java
static final class LoadRecordingSubchannelPicker extends SubchannelPicker {

    private static final ClientStreamTracer NOOP_CLIENT_STREAM_TRACER =
        new ClientStreamTracer() {
        };
    private static final ClientStreamTracer.Factory NOOP_CLIENT_STREAM_TRACER_FACTORY =
        new ClientStreamTracer.Factory() {
          @Override
          public ClientStreamTracer newClientStreamTracer(StreamInfo info, Metadata headers) {
            return NOOP_CLIENT_STREAM_TRACER;
          }
        };

    private final ClientLoadCounter counter;
    private final SubchannelPicker delegate;

    LoadRecordingSubchannelPicker(ClientLoadCounter counter, SubchannelPicker delegate) {
      this.counter = checkNotNull(counter, "counter");
      this.delegate = checkNotNull(delegate, "delegate");
    }

    @Override
    public PickResult pickSubchannel(PickSubchannelArgs args) {
      PickResult result = delegate.pickSubchannel(args);
      if (!result.getStatus().isOk()) {
        return result;
      }
      if (result.getSubchannel() == null) {
        return result;
      }
      ClientStreamTracer.Factory originFactory = result.getStreamTracerFactory();
      if (originFactory == null) {
        originFactory = NOOP_CLIENT_STREAM_TRACER_FACTORY;
      }
      XdsClientLoadRecorder recorder = new XdsClientLoadRecorder(counter, originFactory);
      return PickResult.withSubchannel(result.getSubchannel(), recorder);
    }
  }
```

Users construct this `LoadRecordingSubchannelPicker` to wrap their `SubchannelPicker`.

Reply:
I like your idea. `interceptPickResult()` does start to look weird when `BackendMetricsRecordPicker` is added. Two things I wanted to confirm:

1. You are talking about removing `StatsStore#interceptPickResult()` from the interface, right?
2. `LocalityStoreTest` verifies `StatsStore#interceptPickResult()` is called. If you remove it, I think the alternative testing approach is to verify the type of the picker is `LoadRecordingSubchannelPicker` and `BackendMetricsRecordPicker` wrapped in one way or the other, both with the expected `ClientLoadCounter`. Is that what's in your mind?

Also, `LoadRecordingSubchannelPicker` and `BackendMetricsRecordPicker` share a lot in common. The only difference is they are getting different `ClientStreamTracer.Factory`s. It might simpler to just combine them into one picker, in which you wrap the factory, instead of the picker, twice. Alternatively, maybe define a base class with `abstract ClientStreamTracer.Factory wrapTracerFactory(ClientStreamTracer.Factory original)`. I don't have strong opinion on this. It's up to you.

Reply:
1. Yes, that what I mean.
2. For testing, since we are not able to access any kind of "picker" in the test, as they will all be wrapped by some higher-level pickers (e.g., `DroppablePicker` or `InterLocalityPicker`). So we do not verify picker instances. Instead, we verify the `ClientStreamTracer.Factory` instance from `PickResult#getStreamTracerFactory()` after we let the picker perform a pick. This makes sense as if we are "expecting some logic has been applied to `PickResult`s".